File: load-tests/docs/scripts/prometheus.go
package main

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

type prometheusAPI interface {
	runtimeInfo() (prometheusResponse, error)
	query(string) (prometheusResponse, error)
}

type prometheusClient struct {
	endpoint   string
	headers    http.Header
	httpClient *http.Client
	timeAnchor string
}

type prometheusResponse struct {
	Status string         `json:"status"`
	Data   prometheusData `json:"data"`
}

type prometheusData struct {
	Result []prometheusSeries `json:"result"`
}

type prometheusSeries struct {
	Metric map[string]string `json:"metric"`
	Value  []json.RawMessage `json:"value"`
}

func newPrometheusClient(opts options) (prometheusClient, error) {
	headers, err := authHeaders(opts)
	if err != nil {
		return prometheusClient{}, err
	}
	return prometheusClient{
		endpoint:   strings.TrimRight(opts.endpoint, "/"),
		headers:    headers,
		httpClient: &http.Client{},
		timeAnchor: opts.timeAnchor,
	}, nil
}

func authHeaders(opts options) (http.Header, error) {
	if opts.bearerToken != "" && (opts.basicAuthUser != "" || opts.basicAuthPass != "") {
		return nil, errors.New("--token cannot be combined with --user/--password.")
	}
	if (opts.basicAuthUser == "") != (opts.basicAuthPass == "") {
		return nil, errors.New("--user and --password must be provided together.")
	}

	headers := http.Header{}
	if opts.bearerToken != "" {
		headers.Set("Authorization", "Bearer "+opts.bearerToken)
	}
	if opts.basicAuthUser != "" {
		credentials := opts.basicAuthUser + ":" + opts.basicAuthPass
		headers.Set("Authorization", "Basic "+base64.StdEncoding.EncodeToString([]byte(credentials)))
	}
	return headers, nil
}

func (c prometheusClient) runtimeInfo() (prometheusResponse, error) {
	return c.getJSON(c.endpoint+"/api/v1/status/runtimeinfo", 15*time.Second)
}

func (c prometheusClient) query(query string) (prometheusResponse, error) {
	values := url.Values{"query": []string{query}}
	if c.timeAnchor != "" {
		values.Set("time", c.timeAnchor)
	}
	return c.getJSON(c.endpoint+"/api/v1/query?"+values.Encode(), 30*time.Second)
}

func (c prometheusClient) getJSON(rawURL string, timeout time.Duration) (prometheusResponse, error) {
	request, err := http.NewRequest(http.MethodGet, rawURL, nil)
	if err != nil {
		return prometheusResponse{}, err
	}
	request.Header = c.headers.Clone()

	httpClient := *c.httpClient
	httpClient.Timeout = timeout
	response, err := httpClient.Do(request)
	if err != nil {
		return prometheusResponse{}, err
	}
	defer response.Body.Close()

	if response.StatusCode < 200 || response.StatusCode >= 300 {
		_, _ = io.Copy(io.Discard, response.Body)
		return prometheusResponse{}, fmt.Errorf("Prometheus returned HTTP %d.", response.StatusCode)
	}

	var parsed prometheusResponse
	if err := json.NewDecoder(response.Body).Decode(&parsed); err != nil {
		return prometheusResponse{}, fmt.Errorf("Prometheus returned invalid JSON: %w", err)
	}
	return parsed, nil
}

func sampleString(value json.RawMessage) string {
	var decoded any
	decoder := json.NewDecoder(bytes.NewReader(value))
	decoder.UseNumber()
	if err := decoder.Decode(&decoded); err != nil {
		return ""
	}
	switch typedValue := decoded.(type) {
	case string:
		return typedValue
	case json.Number:
		return typedValue.String()
	default:
		return ""
	}
}

func checkEndpoint(client prometheusAPI, endpoint string) error {
	response, err := client.runtimeInfo()
	if err != nil {
		return errors.New(prometheusEndpointHelp(endpoint))
	}
	if response.Status != "success" {
		return fmt.Errorf("Endpoint '%s' is reachable, but it did not return a Prometheus API success response.\n\nIf you are running locally, make sure the port-forward points at Prometheus:\n\n  kubectl port-forward -n monitoring svc/kube-prometheus-stack-prometheus 9090:9090", endpoint)
	}
	return nil
}

func prometheusEndpointHelp(endpoint string) string {
	return fmt.Sprintf(`Could not reach Prometheus endpoint '%s'.

If you are running locally, start the port-forward in another terminal:

  kubectl port-forward -n monitoring svc/kube-prometheus-stack-prometheus 9090:9090

Then rerun this script with:

  --endpoint http://localhost:9090

If you are using the CI monitor ingress, verify the URL and pass credentials with
--user/--password, for example:

  --endpoint https://ci-monitor.benchmark.camunda.cloud --user "$PROM_USER" --password "$PROM_PASS"`, endpoint)
}

File: load-tests/docs/scripts/load_test_report/src/load_test_report/prometheus.py
"""Prometheus HTTP client and authentication helpers."""

import base64
import json
from collections.abc import Mapping
from typing import Any
from urllib.error import HTTPError
from urllib.error import URLError
from urllib.parse import urlencode
from urllib.request import Request
from urllib.request import urlopen

from .errors import ReportError


class PrometheusClient:
    def __init__(
        self,
        endpoint: str,
        bearer_token: str,
        basic_auth_user: str,
        basic_auth_password: str,
        time_anchor: str,
    ):
        self.endpoint = endpoint.rstrip("/")
        self.headers = auth_headers(bearer_token, basic_auth_user, basic_auth_password)
        self.time_anchor = time_anchor

    def runtime_info(self) -> Mapping[str, Any]:
        return self._get_json(f"{self.endpoint}/api/v1/status/runtimeinfo", timeout=15)

    def query(self, query: str) -> Mapping[str, Any]:
        params = {"query": query}
        if self.time_anchor:
            params["time"] = self.time_anchor
        return self._get_json(f"{self.endpoint}/api/v1/query?{urlencode(params)}")

    def _get_json(self, url: str, timeout: int = 30) -> Mapping[str, Any]:
        request = Request(url, headers=self.headers)
        try:
            with urlopen(request, timeout=timeout) as response:
                parsed = json.loads(response.read().decode("utf-8"))
        except (HTTPError, URLError, TimeoutError, OSError) as error:
            raise ReportError("query failed") from error
        except json.JSONDecodeError as error:
            raise ReportError(f"Prometheus returned invalid JSON: {error}") from error
        if not isinstance(parsed, Mapping):
            raise ReportError("Prometheus returned a non-object JSON response.")
        return parsed


def auth_headers(bearer_token: str, basic_auth_user: str, basic_auth_password: str) -> dict[str, str]:
    if bearer_token and (basic_auth_user or basic_auth_password):
        raise ReportError("--token cannot be combined with --user/--password.")
    if bool(basic_auth_user) != bool(basic_auth_password):
        raise ReportError("--user and --password must be provided together.")
    headers: dict[str, str] = {}
    if bearer_token:
        headers["Authorization"] = f"Bearer {bearer_token}"
    if basic_auth_user:
        credentials = f"{basic_auth_user}:{basic_auth_password}".encode()
        headers["Authorization"] = f"Basic {base64.b64encode(credentials).decode('ascii')}"
    return headers


def check_endpoint(client: PrometheusClient, endpoint: str) -> None:
    try:
        response = client.runtime_info()
    except ReportError as error:
        raise ReportError(prometheus_endpoint_help(endpoint)) from error
    if response.get("status") != "success":
        raise ReportError(
            f"""Endpoint '{endpoint}' is reachable, but it did not return a Prometheus API success response.

If you are running locally, make sure the port-forward points at Prometheus:

  kubectl port-forward -n monitoring svc/kube-prometheus-stack-prometheus 9090:9090"""
        )


def prometheus_endpoint_help(endpoint: str) -> str:
    return f'''Could not reach Prometheus endpoint '{endpoint}'.

If you are running locally, start the port-forward in another terminal:

  kubectl port-forward -n monitoring svc/kube-prometheus-stack-prometheus 9090:9090

Then rerun this script with:

  --endpoint http://localhost:9090

If you are using the CI monitor ingress, verify the URL and pass credentials with
--user/--password, for example:

  --endpoint https://ci-monitor.benchmark.camunda.cloud --user "$PROM_USER" --password "$PROM_PASS"'''
